File: gen3rl/features/encoder.py
from __future__ import annotations
import re
import numpy as np
from .schema import FEATURE_SPECS, TYPE_NAMES, MoveRole, gen3_damage_class, hp_bucket, power_bucket, accuracy_bucket, pp_bucket
from .move_semantics import MoveClass, VARIABLE_DAMAGE_MOVES, classify_move, effectiveness_feature

STATUS_IDS = {"": 0, "brn": 1, "par": 2, "psn": 3, "tox": 3, "slp": 4, "frz": 5}

def parse_condition(text: str) -> tuple[int, int, str]:
    if "fnt" in text: return 0, 1, ""
    match = re.match(r"(\d+)/(\d+)(?:\s+(\w+))?", text or "")
    if not match: return 1, 1, ""
    return int(match.group(1)), int(match.group(2)), match.group(3) or ""

def classify_role(move: dict) -> int:
    mid = str(move.get("id") or move.get("move") or "").lower().replace(" ", "")
    if mid in {"protect", "detect", "endure"}: return MoveRole.PROTECT
    if mid in {"recover", "softboiled", "rest", "synthesis", "moonlight", "morningsun", "slackoff"}: return MoveRole.RECOVERY
    if mid in {"sunnyday", "raindance", "sandstorm", "hail"}: return MoveRole.WEATHER
    if mid in {"spikes"}: return MoveRole.FIELD
    if mid in {"roar", "whirlwind"}: return MoveRole.PHAZE
    if mid in {"batonpass"}: return MoveRole.PIVOT
    if mid in {"explosion", "selfdestruct", "memento"}: return MoveRole.SELF_KO
    if classify_move(move) == MoveClass.FIXED_DAMAGE: return MoveRole.FIXED
    if classify_move(move) == MoveClass.NORMAL_DAMAGE: return MoveRole.DAMAGE
    boosts=move.get("boosts") or {}
    if boosts and any(int(v)<0 for v in boosts.values()): return MoveRole.DEBUFF
    if boosts or move.get("self", {}).get("boosts"): return MoveRole.SETUP
    if move.get("status"): return MoveRole.STATUS
    return MoveRole.UTILITY

def encode_request(request: dict) -> tuple[np.ndarray, np.ndarray]:
    """Encode move candidates using only a player-specific Showdown request."""
    out = np.zeros((9, len(FEATURE_SPECS)), dtype=np.int64)
    mask = np.zeros(9, dtype=bool)
    side = request.get("side") or {}
    mons = side.get("pokemon") or []
    own = next((p for p in mons if p.get("active")), mons[0] if mons else {})
    hp, max_hp, status = parse_condition(own.get("condition", ""))
    public = request.get("public") or {}
    target = public.get("target") or {}
    target_status = target.get("status") or ""
    weather_name = str(public.get("weather", "")).lower()
    weather = 1 if "sun" in weather_name else 2 if "rain" in weather_name else 3 if "sand" in weather_name else 4 if "hail" in weather_name else 0
    own_speed=int((own.get("stats") or {}).get("spe",0)); target_speed=int(target.get("estimatedSpeed",own_speed))
    speed_relation=0 if own_speed<target_speed else 2 if own_speed>target_speed else 1
    moves = ((request.get("active") or [{}])[0] or {}).get("moves", [])
    forced_switch = bool((request.get("forceSwitch") or [False])[0])
    for slot, move in enumerate(moves[:4]):
        move_pp=move.get("pp")
        legal = not forced_switch and not move.get("disabled", False) and (move_pp is None or move_pp > 0)
        mask[slot] = legal
        mtype = str(move.get("type", "unknown")).lower()
        power = int(move.get("basePower", move.get("basePowerCallback", 0)) or 0)
        row = out[slot]
        row[0] = slot; row[1] = int(classify_role(move)); row[2] = TYPE_NAMES.index(mtype) if mtype in TYPE_NAMES else 17
        move_class = classify_move(move)
        try: row[3] = int(gen3_damage_class(mtype, power if move_class == MoveClass.STATUS else max(power, 1)))
        except ValueError: row[3] = 2 if power <= 0 else 0
        row[4] = power_bucket(power); row[5] = accuracy_bucket(move.get("accuracy", True))
        priority = int(move.get("priority", 0)); row[6] = 0 if priority < 0 else 2 if priority > 0 else 1
        row[7] = int(mtype in {str(t).lower() for t in own.get("types", [])})
        row[8] = int(effectiveness_feature(move, target.get("types") or [], target_status))
        row[9] = pp_bucket(int(move_pp or 0))
        row[10] = hp_bucket(hp, max_hp); row[11] = int(target.get("hpBucket", 4)); row[12] = STATUS_IDS.get(status, 0)
        row[13] = STATUS_IDS.get(target_status, 0); row[14] = speed_relation; row[15] = 0; row[16] = 0
        row[17] = int(int(public.get("turn", 0)) <= 1); row[18] = weather; row[19] = 2
    # Switch action indices are universal and stay logically separate from the ROM move policy.
    switches = [p for p in mons if not p.get("active") and "fnt" not in p.get("condition", "")]
    active = (request.get("active") or [{}])[0] or {}
    trapped = bool(active.get("trapped") or active.get("maybeTrapped"))
    if forced_switch or not trapped:
        mask[4:4 + min(5, len(switches))] = True
    if request.get("wait") or request.get("teamPreview"): mask[:] = False
    return out, mask

File: web-playtest/src/policy/encoder.ts
import type {BattleRequest, MoveRequest} from '../types'
import {FEATURE_INDEX, FEATURE_NAMES, FEATURE_VALUES} from './schema'
import {classifyMove, resolveMoveSemantics} from './moveSemantics'

const STATUS_IDS: Record<string, number> = {'': 0, brn: 1, par: 2, psn: 3, tox: 3, slp: 4, frz: 5}
const PHYSICAL = new Set(['normal', 'fighting', 'flying', 'poison', 'ground', 'rock', 'bug', 'ghost', 'steel'])
const SPECIAL = new Set(['fire', 'water', 'grass', 'electric', 'psychic', 'ice', 'dragon', 'dark'])
const ROLE_IDS = Object.fromEntries(FEATURE_VALUES.move_role.map((name, index) => [name, index]))

export function parseCondition(text = ''): [number, number, string] {
  if (text.includes('fnt')) return [0, 1, '']
  const match = text.match(/(\d+)\/(\d+)(?:\s+(\w+))?/)
  return match ? [Number(match[1]), Number(match[2]), match[3] || ''] : [1, 1, '']
}
export const hpBucket = (hp: number, max: number) => hp <= 0 || max <= 0 ? 0 : hp * 4 <= max ? 1 : hp * 2 <= max ? 2 : hp * 4 <= max * 3 ? 3 : 4
const powerBucket = (v: number) => v <= 0 ? 0 : v <= 40 ? 1 : v <= 70 ? 2 : v <= 100 ? 3 : 4
const accuracyBucket = (v: number | boolean | null | undefined) => v === true || v == null ? 0 : Number(v) <= 70 ? 1 : Number(v) <= 85 ? 2 : Number(v) < 100 ? 3 : 4
const ppBucket = (v: number) => v <= 0 ? 0 : v <= 4 ? 1 : v <= 9 ? 2 : v <= 19 ? 3 : 4

export function classifyRole(move: MoveRequest): number {
  const id = String(move.id || move.move || '').toLowerCase().replaceAll(' ', '')
  if (['protect', 'detect', 'endure'].includes(id)) return ROLE_IDS.protect
  if (['recover', 'softboiled', 'rest', 'synthesis', 'moonlight', 'morningsun', 'slackoff'].includes(id)) return ROLE_IDS.recovery
  if (['sunnyday', 'raindance', 'sandstorm', 'hail'].includes(id)) return ROLE_IDS.weather
  if (id === 'spikes') return ROLE_IDS.field
  if (['roar', 'whirlwind'].includes(id)) return ROLE_IDS.phaze
  if (id === 'batonpass') return ROLE_IDS.pivot
  if (['explosion', 'selfdestruct', 'memento'].includes(id)) return ROLE_IDS.self_ko
  if (['seismictoss', 'nightshade', 'dragonrage', 'sonicboom', 'psywave', 'superfang'].includes(id)) return ROLE_IDS.fixed
  if (classifyMove(move) === 'normal-damage') return ROLE_IDS.damage
  if (move.boosts && Object.values(move.boosts).some(value => value < 0)) return ROLE_IDS.debuff
  if (move.boosts || move.self?.boosts) return ROLE_IDS.setup
  if (move.status) return ROLE_IDS.status
  return ROLE_IDS.utility
}

export function encodeRequest(request: BattleRequest): {features: number[][]; mask: boolean[]} {
  const features = Array.from({length: 9}, () => Array(FEATURE_NAMES.length).fill(0))
  const mask = Array(9).fill(false)
  const mons = request.side?.pokemon || []
  const own = mons.find(mon => mon.active) || mons[0] || {condition: ''}
  const [hp, maxHp, status] = parseCondition(own.condition)
  const target = request.public?.target
  const weatherName = (request.public?.weather || '').toLowerCase()
  const weather = weatherName.includes('sun') ? 1 : weatherName.includes('rain') ? 2 : weatherName.includes('sand') ? 3 : weatherName.includes('hail') ? 4 : 0
  const ownSpeed = own.stats?.spe || 0
  const targetSpeed = target?.estimatedSpeed ?? ownSpeed
  const speed = ownSpeed < targetSpeed ? 0 : ownSpeed > targetSpeed ? 2 : 1
  const forced = Boolean(request.forceSwitch?.[0])
  for (const [slot, move] of (request.active?.[0]?.moves || []).slice(0, 4).entries()) {
    const pp = move.pp
    mask[slot] = !forced && !move.disabled && (pp == null || pp > 0)
    const type = (move.type || 'unknown').toLowerCase()
    const power = Number(move.basePower || 0)
    const row = features[slot]
    row[0] = slot; row[1] = classifyRole(move); row[2] = FEATURE_VALUES.move_type.indexOf(type)
    if (row[2] < 0) row[2] = 17
    const moveClass = classifyMove(move)
    row[3] = moveClass === 'status' ? 2 : PHYSICAL.has(type) ? 0 : SPECIAL.has(type) ? 1 : 0
    row[4] = powerBucket(power); row[5] = accuracyBucket(move.accuracy)
    row[6] = (move.priority || 0) < 0 ? 0 : (move.priority || 0) > 0 ? 2 : 1
    row[7] = Number((own.types || []).some(value => value.toLowerCase() === type))
    row[8] = resolveMoveSemantics(move, target?.types || [], target?.status || '').effectivenessBucket
    row[9] = ppBucket(Number(pp || 0))
    row[10] = hpBucket(hp, maxHp); row[11] = Number(target?.hpBucket ?? 4)
    row[12] = STATUS_IDS[status] || 0; row[13] = STATUS_IDS[target?.status || ''] || 0
    row[14] = speed; row[15] = 0; row[16] = 0; row[17] = Number((request.public?.turn || 0) <= 1)
    row[18] = weather; row[19] = 2
  }
  const switches = mons.filter(mon => !mon.active && !mon.condition.includes('fnt'))
  const trapped = Boolean(request.active?.[0]?.trapped)
  if (forced || !trapped) switches.slice(0, 5).forEach((_, index) => { mask[4 + index] = true })
  return {features, mask}
}

export function featureCategories(row: number[]): Record<string, string> {
  return Object.fromEntries(FEATURE_NAMES.map((name, index) => [name, FEATURE_VALUES[name][row[index]]]))
}
